### PR TITLE
feat: optional rustls support

### PR DESCRIPTION
**Pull request by sunmy2019**
This is the initial implementation of rustls support.

Use it by disabling `native-tls-support` flag and enabling `rustls-support` flag.

Notes:
1. It passes all tests. Don't know if that's enough.
2. There is no available package other than openssl that can handle all pkcs12 pbe methods. The only available create is `p12` that handles some of the PBE methods. I choose one of them as you can see in `examples/tls/create_self_signed_cert.sh`
3. features flags like: "rustls-support" -> "tls-support", "native-tls-support" -> "tls-support".

Feedback is welcome. I am new to this project.

### CODE REVIEW

**Comment by sunmy2019** [comment]
Seem the check requires it passes on both `native-tls` and `rustls` support enabled.

Supporting both enabled is not common in other crates. I deliberately disabled this.

What should we do here?

**Comment by rapiz1** [comment]
https://github.com/taiki-e/cargo-hack has some flags to adjust the powerset behaviour

File: Cargo.toml [hunk 1]
```diff
@@ -11,18 +11,39 @@ build = "build.rs"
...
+]
 
 # Run as a server
 server = []
 # Run as a client
 client = []
+
 # TLS support
-tls = ["tokio-native-tls"]
+tls-support = []
```
Comment (rapiz1) [inline Cargo.toml:29]: what is this used for?

File: Cargo.toml [hunk 2]
```diff
@@ -11,18 +11,39 @@ build = "build.rs"
...
+native-tls-support = ["tokio-native-tls", "tls-support"]
```
Comment (rapiz1) [inline Cargo.toml:30]: I would suggest removing `-support`

File: Cargo.toml (hunk 1 above)
Comment (sunmy2019) [reply Cargo.toml:29]: This enables the common part of both "native-tls" and "rustls".

The use of this feature can be replaced by `#cfg[any(feature="native-tls", feature="rustls")]` in all places. Which is better?

File: Cargo.toml (hunk 2 above)
Comment (sunmy2019) [reply Cargo.toml:30]: I will fix it.

**Comment by rapiz1** [comment]
> It passes all tests. Don't know if that's enough.

Did you check how will the integration test run with this new compile flag? I think we're only running with the default tls implementation ( native-tls ). If we run the test with both native-tls and rustls, we can be sure it's correct.

**Comment by sunmy2019** [comment]
> Did you check how will the integration test run with this new compile flag? I think we're only running with the default tls implementation ( native-tls )

I run it locally by replacing native-tls with rustls.

I am figuring out how to test both.

**Comment by sunmy2019** [comment]
I added both tests to CI.

File: src/main.rs [hunk 3]
```diff
@@ -2,6 +2,7 @@ use anyhow::Result;
 use clap::Parser;
 use rathole::{run, Cli};
 use tokio::{signal, sync::broadcast};
+#[cfg(not(feature = "console"))]
```
Comment (sunmy2019) [inline src/main.rs:5]:
```suggestion
```

File: src/transport/websocket.rs [hunk 4]
```diff
@@ -30,7 +34,8 @@ impl TransportStream {
     fn get_tcpstream(&self) -> &TcpStream {
         match self {
             TransportStream::Insecure(s) => s,
-            TransportStream::Secure(s) => s.get_ref().get_ref().get_ref(),
+
```
Comment (sunmy2019) [inline src/transport/websocket.rs:37]:
```suggestion
```

File: Cargo.toml [hunk 5]
```diff
@@ -11,18 +11,52 @@ build = "build.rs"
...
 # TLS support
-tls = ["tokio-native-tls"]
+native-tls = ["tokio-native-tls"]
+rustls = [
+    "tokio-rustls",
+    "rustls-pemfile",
+    "rustls-native-certs",
+    "p12",
+]
+# original tls feature for backward compatibility
```
Comment (rapiz1) [inline Cargo.toml:36]: For backward compatibility of what? Can we get rid of it?

File: src/lib.rs [hunk 6]
```diff
@@ -119,8 +119,8 @@ async fn run_instance(
     args: Cli,
     shutdown_rx: broadcast::Receiver<bool>,
     service_update: mpsc::Receiver<ConfigChange>,
-) {
```
Comment (rapiz1) [inline src/lib.rs:122]: 👍

File: src/transport/tls.rs [hunk 7]
```diff
@@ -1,111 +1,296 @@
...
-                    .build()?;
-                Some(TlsConnector::from(connector))
-            }
-            None => {
-                // if no trusted_root is specified, allow TlsConnector to use system default
-                let connector = native_tls::TlsConnector::builder().build()?;
-                Some(TlsConnector::from(connector))
-            }
-        };
+#[cfg(feature = "native-tls")]
```
Comment (rapiz1) [inline src/transport/tls.rs:4]: is it possible to split two tls implementation into two files? That will definitely look much better. For example, `native_tls.rs` and `rustls.rs`, and we import and alias one of them as the final tls transport in `tls.rs`. Even in subdirectory `tls`, if that's more convenient.

File: Cargo.toml (hunk 5 above)
Comment (sunmy2019) [reply Cargo.toml:36]: Yes. For backward compatibility of original feature gate.

File: src/transport/tls.rs (hunk 7 above)
Comment (sunmy2019) [reply src/transport/tls.rs:4]: Done

File: Cargo.toml (hunk 5 above)
Comment (sunmy2019) [reply Cargo.toml:36]: Removed

File: tests/integration_test.rs [hunk 8]
```diff
@@ -94,24 +94,29 @@ async fn udp() -> Result<()> {
...
     test("tests/for_udp/websocket_tls_transport.toml", Type::Udp).await?;
 
     Ok(())
 }
 
 #[instrument]
 async fn test(config_path: &'static str, t: Type) -> Result<()> {
+    if cfg!(not(all(feature = "client", feature = "server"))) {
+        // Skip the test if the client or the server is not enabled
+        return Ok(());
```
Comment (rapiz1) [inline tests/integration_test.rs:117]: Why we didn't need this before?

**Review by rapiz1 — APPROVED** [review]
Well done and thank you very much! Here's my last question

File: tests/integration_test.rs (hunk 8 above)
Comment (sunmy2019) [reply tests/integration_test.rs:117]: This combinations are never run in CI before.

I ran `cargo hack test` locally and met this.

**Comment by sunmy2019** [comment]
I will add docs in the sequential PR.

**Comment by skbeh** [comment]
@sunmy2019 The `p12` crate is incompatible with pkcs12 cert generated by OpenSSL3: https://github.com/hjiayz/p12/issues/13.

**Comment by sunmy2019** [comment]
> @sunmy2019 The `p12` crate is incompatible with pkcs12 cert generated by OpenSSL3: [hjiayz/p12#13](https://github.com/hjiayz/p12/issues/13).

See my updated docs here: #337

One difference is that, the crate we use for loading PKCS#12 archives can only handle limited types of PBE algorithms. We only support PKCS#12 archives that they (crate `p12`) support. So we need to specify the PBE algorithm when creating the PKCS#12 archive.

In short, the command to create the PKCS#12 archive with `rustls` support is:

```sh
openssl pkcs12 -export -out identity.pfx -inkey server.key -in server.crt -certfile ca_chain_certs.crt \
    -keypbe PBE-SHA1-3DES -certpbe PBE-SHA1-3DES
```

**Comment by sunmy2019** [comment]
> @sunmy2019 Both SHA1 and 3DES are not secure according to today's security standards. IMHO, a more updated crate should be used to handle modern algorithms.

Do you have suggestions?

**Comment by sunmy2019** [comment]
> @sunmy2019 Both SHA1 and 3DES are not secure according to today's security standards. IMHO, a more updated crate should be used to handle modern algorithms.

That makes sense. I think we should support the default PBEs for newer formats, but I am unsure how.

**Comment by skbeh** [comment]
@sunmy2019 I think we would better to deprecate PKCS#12 since it isn't used very widely and its encryption is poor supported.

**Comment by skbeh** [comment]
According to https://stackoverflow.com/a/62863490/7878845, PKCS#12 file can be generated without encryption, which could be an alternative to not provide false sense of security.

**Comment by sunmy2019** [comment]
> According to https://stackoverflow.com/a/62863490/7878845, PKCS#12 file can be generated without encryption, which could be an alternative to not provide false sense of security.

I've been wondering what's the point of this password in this program when you write down the password in a separate file.

**Comment by skbeh** [comment]
@rapiz1 Also please update the document to use 3DES instead of RC2 (i.e. `-keypbe PBE-SHA1-3DES -certpbe PBE-SHA1-3DES` since 3DES provides not good but still acceptable encryption if someone ever wants it.

**Pull request merged**